### DISCUSSION TITLE
What made you start this project?

### DISCUSSION THREAD

**Discussion by kleenkanteen** [SpartanJ/ecode#91] [General]
@SpartanJ @lupuionut I read the readme, searched through issues and discussions quickly for this question, but no hits. I have 2:

1. What made you take the monumental task of making a whole desktop app framework? 
2. How does it compare to Flutter or React Native or MAUI for cross platform frameworks?

**Comment by SpartanJ** [comment]
Hi @kleenkanteen !
1. I don't think there's an answer. I love to program and it's my hobby. I started working on [eepp](https://github.com/SpartanJ/eepp) many years ago, just as something to pass time and learn, and I continued adding features, step by step until it became the framework that today it is. I used the framework for several private mobile applications and it worked well enough to continue using it, the simplicity of compiling the application natively on my main OS and exporting it to Android was too convenient (easy debugging, faster testing, faster iteration). Using it for work helped a lot to mature the framework. But honestly, the thing is that I simply put much time/dedication into it, it's my learning material, for me is like writing a book or doing art. If I want to know how something works, I read about it, implement it, and repeat that process indefinitely.
ecode was the result of another experiment, one day thought it would be cool to implement a basic text editor to be able to edit the ecode UI XML files in a UI editor, and I kind of got really engaged with the idea of improving the text editor and I ended up starting working on a general purpose code editor. One of the main motivations was that then, I wasn't happy with any of the general-purpose code editors I was using for my casual editing (PHP, JavaScript, Python, and several scripting languages, my most used editor was Geany). None of the editors felt "right" to me, and I didn't need too many features, mainly some good linting, simple auto-completion, and syntax highlighting. And given that I was motivated I kept implementing new features until I found myself using the editor daily for most of my editing stuff. I think that right now there are several interesting simple yet very powerful general-purpose editors (Lapce, Kate, Lite-XL, Zed), but at the time I started working most of them even exists (except for Kate, which existed for many years). Anyway none of them feel "right" to me, I designed ecode to fit my needs. Usually, the issues I have with most editors are very subtle and won't matter to most users.

2. I think I don't know enough about any of those frameworks to actually make a valuable comment without obvious errors. But I can mention a few things (that I might be wrong or misinformed):
eepp GUI was designed with inspiration from two main technologies: HTML+CSS and Android UI and designed to be 100% platform-independent from the start, this separates the design a little bit from those three frameworks (Flutter and React Native were mainly mobile-oriented, and regarding MAUI I literally know nothing about it, I just discovered with your comment, but I can't find much information). eepp GUI is closer to Qt than those libraries. eepp GUI it's a retained-mode UI with separations of concerns (XML for layout, CSS for style, and the logic is coded). It's very much a work in progress, many features aren't available yet and need a lot of polishing, but it's a good shape to already create very rich interfaces (and many features aren't easily visible from ecode). [here](https://eepp.ensoft.dev/page_uiintroduction.html) is a very short introduction to how it works if you are interested. Good documentation is still lacking, but you know, I'm the only person working on both projects right now, so I kind of prefer to focus on coding than documentation (I have writing documentation haha).

I'm pretty sure I didn't answer your questions but keep asking if you want to. I'll try to answer as clearly as I can.

**Reply by kleenkanteen** [reply]
@SpartanJ Just got around to reading your answer. Very interesting. A few more questions:

1. At work, how did you convince your manager to use your framework?
2. You said if you want to know how something works, you implement it and repeat that process. Do you also look at how other people did the same thing and compare with your approach?
3. How old are you and what do you do for work? (you can give an age range if you don't want to give a number)
4. Have you thought of contributing to a more widely used open source project? You seem to have got a lot of knowledge and experience from this project. That may be better used elsewhere towards a product with a lot of users.

**Comment by SpartanJ** [comment]
Hi again @kleenkanteen !
1. I was an independent developer, I was my own boss, so I made all the decisions. Basically, I worked doing some contract work that was 100% controlled by me.
2. Absolutely. All the time! It's part of the fun too, a lot of code has been inspired by other people's implementations. Usually, I comment where I took inspiration on the code, and also, it's in the [README - Acknowledgements](https://github.com/SpartanJ/ecode).
3. 36 yo, started programming around 14-15. Ex-CTO in a telemedicine company, and now very recently working as a developer in a spreadsheet+docs private alternative software (software not released yet, so I can't disclose it).
4. Not really, I enjoy the process of working from scratch by myself, that's how I like it. I made a few contributions here and there, but nothing significant. I guess I care about having an impact on the commercial products that I work on, I already had a direct impact on millions of people, so I kind of have enough of that. Hobby projects are very different, you can do whatever you want whenever you want it, and that cannot be done in professional environments.

**Reply by kleenkanteen** [reply]
@SpartanJ Interesting. A few more questions:

1. How do you work on your projects on the side? Do you have a schedule for after work? 
2. How did you have an impact on millions of people exactly?
3. What do you like particularly about graphics programming?
4. Are you planning on retiring early so you can dedicate yourself full time to making what you want?

**Comment by SpartanJ** [comment]
Sorry for the delay haha

> How do you work on your projects on the side? Do you have a schedule for after work?

No schedule whatsoever, basically when I have some free time, instead of playing games or looking some movie I like to code to relax sometimes. Usually on weekends, and some nights over the week if I'm not too tired. This is something that varies a lot, there are months that I have time and energy and others I can barely touch any code, I don't pressure myself, this is done for fun.

> How did you have an impact on millions of people exactly?

The platform I worked on is for online medical consultations between patients and doctors, the service has been used extensively over the pandemic, having great positive impact in peoples life's.

> What do you like particularly about graphics programming?

I'm still not sure if I actually like it, it's a means to an end. But the experience was always painful to me, it was rewarding when some of the concepts finally clicked but I'm no expert by any means.

> Are you planning on retiring early so you can dedicate yourself full time to making what you want?

The short answer is yes. But it's more like a dream than a plan, I'm not in a position where I can be certain that I'll achieve that goal (I guess no-one can without knowing the future). Anyway, I'm working in stuffs that I love, so I can't complain.
